include/oneapi/dpl/pstl/hetero/dpcpp/sycl_traits.h

```diff
@@ -24,24 +24,23 @@
 #ifndef _ONEDPL_SYCL_TRAITS_H
 #define _ONEDPL_SYCL_TRAITS_H
 
-#if __INTEL_LLVM_COMPILER && (__INTEL_LLVM_COMPILER < 20240100)
-
-#    define _ONEDPL_DEVICE_COPYABLE(TYPE)                                                                              \
-        template <typename... Ts>                                                                                      \
-        struct sycl::is_device_copyable<TYPE<Ts...>, ::std::enable_if_t<!std::is_trivially_copyable_v<TYPE<Ts...>>>>   \
-            : ::std::conjunction<sycl::is_device_copyable<Ts>...>                                                      \
-        {                                                                                                              \
-        };
-
+#if _ONEDPL_SYCL_DEVICE_COPYABLE_SPECIALIZATION_BROKEN
+// Prior to the particular version of SYCL library implementation, sycl::is_device_copyable relied upon a second
+// template parameter to resolve ambiguity with the general is_trivially_copyable trait. This does not follow the SYCL
+// specification, but is required to support older versions of SYCL library. Without this workaround, the user will face a
+// compile time error of ambiguous partial template specialization.
+#   define _ONEDPL_SPECIALIZE_FOR(TYPE, ...) TYPE<__VA_ARGS__>, std::enable_if_t<!std::is_trivially_copyable_v<TYPE<__VA_ARGS__>>>
 #else
+#   define _ONEDPL_SPECIALIZE_FOR(TYPE, ...) TYPE<__VA_ARGS__>
+#endif //__INTEL_LLVM_COMPILER && (__INTEL_LLVM_COMPILER < 20240100)
+
+namespace
+{
 
-#    define _ONEDPL_DEVICE_COPYABLE(TYPE)                                                                              \
-        template <typename... Ts>                                                                                      \
-        struct sycl::is_device_copyable<TYPE<Ts...>> : ::std::conjunction<sycl::is_device_copyable<Ts>...>             \
-        {                                                                                                              \
-        };
+template <typename... _Ts>
+struct __is_device_copyable: std::conjunction<sycl::is_device_copyable<_Ts>...> {};
```

Comment:
Maybe rename to `__device_copyable`, which would better read as both "singular" and "plural"? It's purely subjective, feel free to ignore.

Reply:
I agree that such a name change is useful.
another option I was thinking could be `__are_all_device_copyable`.
I'm not sure which I like better.  `__device_copyable` makes singular and plural more understandable, but we perhaps lose the idea that we are checking a trait.

Reply:
I've updated to `__are_all_device_copyable`, though I'm not tied to it, willing to change if someone has an argument.